stage0/src/acpi_tables.rs

```diff
@@ -184,6 +192,73 @@ impl DescriptionHeader {
     }
 }
 
+/// Root System Description Table.
+///
+/// See Section 5.2.7 in the ACPI specification for more details.
+#[repr(C, packed)]
+pub struct Rsdt {
+    header: DescriptionHeader,
+    // The RSDT contains an array of pointers to other tables, but unfortunately this can't be
+    // expressed in Rust.
+}
+
+impl Rsdt {
+    const SIGNATURE: &[u8; 4] = b"RSDT";
+
+    pub fn new(addr: VirtAddr) -> Result<&'static Rsdt, &'static str> {
+        // Safety: we're checking that it's a valid XSDT in `validate()`.
+        let rsdt = unsafe { &*addr.as_ptr::<Rsdt>() };
+        rsdt.validate()?;
+        Ok(rsdt)
+    }
+
+    fn validate(&self) -> Result<(), &'static str> {
+        self.header.validate()?;
+
+        if self.header.signature != *Self::SIGNATURE {
+            return Err("Invalid signature for RSDT table");
+        }
+
+        if (self.header.length as usize - size_of::<DescriptionHeader>()) % size_of::<u32>() != 0 {
+            return Err("RSDT invalid: entries size not a multiple of pointer size");
+        }
+
+        // Safety: we're never dereferencing the pointer.
+        let ebda_base = unsafe { EBDA.as_ptr() } as usize;
+        for &entry in self.entries() {
+            let ptr: usize = entry as usize;
+            if !(ebda_base..ebda_base + EBDA_SIZE).contains(&ptr) {
+                return Err("RSDT invalid: entry points outside EBDA");
+            }
+        }
+        Ok(())
+    }
+
+    fn entries(&self) -> &[u32] {
+        let entries_base = self as *const _ as usize + size_of::<DescriptionHeader>();
+        // Safety: we've validated that the address and length makes sense in `validate()`.
+        unsafe {
+            slice::from_raw_parts(
+                entries_base as *const u32,
+                (self.header.length as usize - size_of::<DescriptionHeader>()) / size_of::<u32>(),
+            )
+        }
+    }
+
+    /// Finds a table based on the signature, if it is present.
+    pub fn get(&self, table: &[u8; 4]) -> Option<&'static DescriptionHeader> {
+        for &entry in self.entries() {
+            let ptr: usize = entry as usize;
+            // Safety: we've checked in `validate()` that all the pointers point within the EBDA.
+            let entry: &'static DescriptionHeader = unsafe { &*(ptr as *const DescriptionHeader) };
```

Comment:
since this is a public struct & function: `validate` is only called in the `new` method. Technically it would be possible to invoke this method without having gone through `validate` in the `new` method.

probs fine tho

Reply:
If you want to use the `DescriptionHeader` as anything else (like `Madt`) that'll call `validate()`.

I'm not entirely happy with how the code is structured right now either (especially around validation), and may refactor it a bit in the future.